km/load_elf.c

```diff
@@ -230,25 +239,40 @@ uint64_t km_load_elf(km_elf_t* e)
          km_guest.km_dynamic_len = phdr->p_filesz;
       }
    }
+   km_infox(KM_TRACE_LOAD, "file=%s min_vaddr=0x%lx ", km_guest.km_filename,
+            km_guest.km_min_vaddr);
 
    km_gva_t adjust = 0;
    /*
-    * Tell static vs dynamic executable.
+    * If the type of the execuatble is ET_EXEC, then the virtual addresses in the ELF must
+    * be honored when loaded because there are literal non-relocatable addresses embedded
+    * in the file.
     *
-    * DYNAMIC section indicates dynamically linked executable or static PIE. The later also has
-    * PT_INTERPRETER.
+    * If the type is ET_DYN, then the file is fully relocatable and we are free to place it
+    * anywhere we want. The further distinction with ET_DYN is whether or not it contains a
+    * PT_INTERP PHDR. If it does then it is an execuatable that uses a dynamic linker and we
+    * load both the program and the dynamic linker specified in the PR_INTERP section.
+    * If there is no PT_INTERP section, we consider this program a dynamic linker itself and
+    * only load it. (Note: a static-PIE looks exactly like a dynamic linker to the ELF loader) 
     *
-    * We load pure static (no DYNAMIC) at the addresses in ELF file, i.e. adjust == 0. For our own
-    * .km files they will load starting at GUEST_MEM_START_VA because its the way we build them.
-    * Others will go where they request, typical Linux exec would start at 4MB.
+    * Default Linux ET_EXEC files start at virtual address 4MB by convention.
+    * We load pure static (no DYNAMIC) at the addresses in ELF file, i.e. adjust == 0. 
+    * For our own .km files they will load starting at GUEST_MEM_START_VA because of
+    * the way we build them.
     *
     * We load DYNAMIC starting at GUEST_MEM_START_VA, adjust will tell how much shift there was. For
     * loading PIE that doesn't matter, but we need to keep it for coredump and such. For dynamic
     * linking adjust is passed to the dynamic linker so it know how to do relocations.
     */
-   if (km_guest.km_dynamic_vaddr != 0 && km_guest.km_dynamic_len != 0) {
+   if (e->ehdr.e_type == ET_EXEC) {
+      km_infox(KM_TRACE_LOAD, "type is ET_EXEC");
+   } else if (e->ehdr.e_type == ET_DYN) {
+      km_infox(KM_TRACE_LOAD, "type is ET_DYN");
```

Comment:
Looking at my system (Fedora 35) I see `/bin/cat` being PIE and `/bin/gcc` being regular dynamic exec:

/bin/gcc: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=1afd7874e01ce0a1ecca42cc3a42a022127e8d50, for GNU/Linux 3.2.0, stripped
/bin/cat: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=b5f22ffbc60b2993dc4f138c5a052e994d404cb7, for GNU/Linux 3.2.0, stripped

However I see them both showing only DYNAMIC header. In fact I was looking for something with ET_EXEC and found none. What am I missing?

With that in mind, I don't have a clear picture how this covers all cases: static, regular dynamic, dynamic PIE, and static PIE.

Reply:
Here is a comment I found in linux `fs/binfmt_elf.c` which I found helpful:
```
 * There are effectively two types of ET_DYN
 * binaries: programs (i.e. PIE: ET_DYN with INTERP)
 * and loaders (ET_DYN without INTERP, since they
 * _are_ the ELF interpreter). The loaders must
 * be loaded away from programs since the program
 * may otherwise collide with the loader (especially
 * for ET_EXEC which does not have a randomized
 * position).
```